relay-server/src/actors/outcome.rs

```diff
@@ -296,9 +316,60 @@ impl Message for TrackRawOutcome {
     type Result = Result<(), OutcomeError>;
 }
 
+/// Common implementation for both processing and non-processing
+impl OutcomeProducer {
+    fn send_batch(&mut self, context: &mut Context<Self>) {
+        log::trace!("sending outcome batch");
+
+        //the future should be either canceled (if we are called with a full batch)
+        // or already called (if we are called by a timeout)
+        self.send_outcomes_future = None;
+
+        if self.unsent_outcomes.is_empty() {
+            log::warn!("unexpected send_batch scheduled with no outcomes to send.");
+            return;
+        }
+
+        let request = SendOutcomes {
+            outcomes: self.unsent_outcomes.drain(..).collect(),
+        };
+
+        self.upstream
+            .send(SendQuery(request))
+            .map(|_| {
+                log::trace!("outcome batch sent.");
+            })
+            .map_err(|error| {
+                log::error!("outcome batch sending failed with: {}", LogError(&error));
+            })
+            .into_actor(self)
+            .spawn(context);
+    }
+
+    fn send_http_message(
+        &mut self,
+        message: TrackRawOutcome,
+        context: &mut Context<Self>,
+    ) -> Result<(), OutcomeError> {
+        log::trace!("Batching outcome");
+        self.unsent_outcomes.push(message);
+        if self.unsent_outcomes.len() >= self.config.max_outcome_batch_size() {
+            if let Some(send_outcomes_future) = self.send_outcomes_future {
+                context.cancel_future(send_outcomes_future);
```

Comment:
I am not 100% sure if this may cancel a HTTP request mid-flight and run the risk of double-sending outcomes. Even if this is an actor future, such a future only has exclusive access to `self` during a poll, not during the lifetime of the future.

Also doesn't this generally cause us to drop a preceding batch of outcomes if we run into the batch size limit again? Seems like that would happen instantly under sustained traffic.

Reply:
The future is removed from the struct the moment that it is started. `self.send_outcomes_future` is only there to synchronize between the batch limit and the batch delay. I think for that reason, neither of the two cases can happen. Can you double-check?

Reply:
I see, so `self.send_outcomes_future` is only being `Some` if we're counting down from the timeout, not when the request is in flight. That makes sense and you're right. Can we somehow make it explicit that nothing that may suspend the future (such as more IO) may be added before line 326?

Reply:
Additionally we can also rename the field to something like `pending_flush` to make this more clear?

Reply:
ok,  changed `send_outcomes_future` to `pending_flush_handle`.

Didn't quite get what you are asking at:
> Can we somehow make it explicit that nothing that may suspend the future (such as more IO) may be added before line 326?

Reply:
I think he meant to write into the comment that nothing should go before the check of `pending_flush_handle` in `send_batch`